content/v2.0/reference/storage-engine.md

```diff
@@ -0,0 +1,157 @@
+---
+title: InfluxDB storage engine
+description: >
+  An overview of the InfluxDB storage engine architecture.
+weight: 7
+menu:
+  v2_0_ref:
+    name: Storage engine
+v2.0/tags: [storage engine, internals, platform]
+---
+
+The InfluxDB storage engine ensures the following three things:
+
+- Data is safely written to disk
+- Queried data is returned complete and correct
+- Data is accurate (first) and performant (second)
+
+This document details the internal workings of the storage engine.
+This information is presented both as a reference and to aid those looking to maximize performance.
+
+Major topics include:
+
+* [Write Ahead Log (WAL)](#write-ahead-log-wal)
+* [Time-Structed Merge Tree (TSM)](#time-structured-merge-tree-tsm)
+* [Time Series Index (TSI)](#time-series-index-tsi)
+
+## Writing data: from API to disk
+
+The storage engine handles data from the point an API request is received through writing it to the physical disk.
+Data is written to InfluxDB using [line protocol](/v2.0/reference/line-) sent via HTTP POST request to the `/write` endpoint.
+Batches of [points](/v2.0/reference/glossary/#point) are sent to InfluxDB, compressed, and written to a WAL for immediate durability.
+The points are also written to an in-memory cache and become immediately queryable.
+The cache is periodically written to disk in the form of [TSM](#time-structured-merge-tree-tsm) files.
+As TSM files accumulate, they are combined and compacted into higher level TSM files.
+
+Points can be sent individually; however, for efficiency, most applications send points in batches.
+A typical batch ranges in size from hundreds to thousands of points.
+Points in a POST body can be from an arbitrary number of series, measurements, and tag sets.
+Points in a batch do not have to be from the same measurement or tagset.
+
+## Write Ahead Log (WAL)
+
+The Write Ahead Log (WAL) ensures durability by retaining data when the storage engine restarts.
+It ensures that written data does not disappear in an unexpected failure.
+When a client sends a write request, the following occurs:
+
+1. Write request is appended to the end of the WAL file.
+2. `fsync()` the data to the file.
+3. Update the in-memory database.
+   <!-- 3. Update the in-memory CACHE? -->
+4. Return success to caller.
+
+`fsync()` takes the file and pushes pending writes all the way through any buffers and caches to disk.
+As a system call, `fsync()` has a kernel context switch which is expensive _in terms of time_ but guarantees your data is safe on disk.
+
+{{% note%}}
+To `fsync()` less frequently, batch your points (send in ~2000 points at a time).
```

Comment:
I don't think we should call out the number of points here. 

1. We don't have any thorough benchmarking done for 2.0 yet.
2. It can change drastically depending on the underlying hardware.

We have recommended batch sizes elsewhere, but those are under review as well.

Reply:
Also, in a related discussion, "points" may not be the write term here. This is something we still need to sort out with the SEs and engineers. A single line of line protocol can contain multiple points. For example:

```
measuremenName,tag1=tag1Val,tag2=tag2Val field1="something,field2="somethingelse",field3=3i 9999999999999
```

This ^ line of line protocol actually contains 3 points. They all share the same measurement, tag set, and time stamp, but there are 3 different fields.

So when we talk about batch size, are we talking about the number of **points** or the number of **lines** of line protocol? I don't know.